### PR TITLE
feat: Logins log for Fail2Ban

### LINKED ISSUE #985: Enhancement: [Failed] Logins log for Fail2Ban
### Contact Details

_No response_

### What features would you like to see added?

A function that writes all logins or at least all failed logins to a file or to the console log.

### More details

Since it would be great especially for Fail2Ban, I think that would be quite a suitable log format:

`[Mon Feb 17, 2023 18:25] [Login] [Login failed] [Username: <Username>] [Request-IP: <IP>]`

For Fail2Ban only the IP is really important, but I think it would be nice to have the whole thing a bit more detailed.

For anyone from the future hoping to find a Fail2Ban regex here, I think this one works:
`^\[.*\] \[Login\] \[Login failed\] \[.*\] \[Request-IP: <HOST>\]`

### Which components are impacted by your request?

General, Other

### Pictures

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by berry-13**
I'm not sure if this is the function that was requested, which is why I welcome ideas on how to improve it. Every time a login fails, the logs are both displayed in the console and saved in the root folder in a file named login-logs.txt #985 

logs:

```
{"level":"error","message":"Passport Local Strategy - Password does not match","parameters":[{"name":"isMatch","value":false}],"timestamp":"2023-09-23T23:24:38.786Z"}
{"level":"error","message":"[Login] [Login failed] [Username: myemail@test.com] [Request-IP: 45.xx.xx.238]","timestamp":"2023-09-23T23:24:38.786Z"}
```

## Change Type

- [x] New feature (non-breaking change which adds functionality)

## Testing

check if the log and the .txt works properly

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes
- [x] Any changes dependent on mine have been merged and published in downstream modules.

### CODE REVIEW

File: api/strategies/localStrategy.js [hunk 1]
```diff
@@ -1,7 +1,17 @@
 const { Strategy: PassportLocalStrategy } = require('passport-local');
 const User = require('../models/User');
 const { loginSchema, errorsToString } = require('./validators');
-const DebugControl = require('../utils/debug.js');
+const winston = require('winston');
```
Comment (danny-avila) [inline api/strategies/localStrategy.js]: It's best practice to import the logger from a central place

File: package.json [hunk 2]
```diff
@@ -93,5 +93,8 @@
       "admin/",
       "packages/"
     ]
+  },
+  "dependencies": {
+    "winston": "^3.10.0"
```
Comment (danny-avila) [inline package.json]: wrong installation location. you will have to do the following steps

```bash
npm uninstall winston  # from root
# your changes to both package and package-lock 
# should reset to what they were before
# OR
# you can do git reset HEAD~3, discard changes to these files,
# and run npm run reinstall to go back to the previous state
# and then 
cd ./api
npm i winston
```

**Review by danny-avila — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (danny-avila) [reply package.json]: make sure to run `npm run reinstall` in either case to make sure package-lock isn't pointing to your cache.

File: api/strategies/localStrategy.js (hunk 1 above)
Comment (berry-13) [reply api/strategies/localStrategy.js]: I'm not sure if I've done what you requested, and I'm not even sure how api\utils\LoggingSystem.js works, but from the tests I've conducted, everything is functioning correctly without any new errors, so I believe it should be fine

Comment (danny-avila) [reply api/strategies/localStrategy.js]: You can just import directly from logger.js no need to import from logging system. That file could probably be deleted

**Review by danny-avila — APPROVED** [review]

**Comment by XHyperDEVX** [comment]
it works, but... the log works if you enter wrong logical data. e.g. username: 123@456.com password: 123456789
but with wrong non logical data, like username: 123 password: 123 [which doesn't work because the username has to be an email and the password has to be min. 8 characters long] it won't be logged! i think that would be important to add.

File: api/strategies/localStrategy.js [hunk 3]
```diff
@@ -1,7 +1,17 @@
...
+
+// Configure Winston logger
+const logger = winston.createLogger({
+  level: 'info',
+  format: winston.format.combine(winston.format.timestamp(), winston.format.json()),
+  transports: [
+    new winston.transports.Console(),
+    new winston.transports.File({ filename: 'login-logs.log' }),
```
Comment (XHyperDEVX) [inline api/strategies/localStrategy.js]: here e.g. /logins/login-logs.txt instead of just login-logs.txt
-> for better handling when using logrotate

**Comment by Berry-13** [comment]
I know, I did it on purpose this way! I thought the logs would help you understand if someone wanted to attempt to brute force the password, etc..

**Comment by XHyperDEVX** [comment]
> I know, I did it on purpose this way! I thought the logs would help you understand if someone wanted to attempt to brute force the password, etc..

No, the wish was to make librechat compatible with fail2ban xD

**Comment by danny-avila** [comment]
> it works, but... the log works if you enter wrong logical data. e.g. username: [123@456.com](mailto:123@456.com) password: 123456789 but with wrong non logical data, like username: 123 password: 123 [which doesn't work because the username has to be an email and the password has to be min. 8 characters long] it won't be logged! i think that would be important to add.

this is a good point, it should count as a login fail when the validation fails @Berry-13

**Pull request merged**